content/blog/2019-10-19-loop-invariant-code-motion/index.md

```diff
@@ -0,0 +1,297 @@
+"""
+[[extra.authors]]
+name = "Rolph Recto"
+link = "https://twitter.com/rolphrecto"
+[[extra.authors]]
+name = "Gabriela Calinao Correa"
+link = "https://twitter.com/elalaCorrea"
++++
+
+
+
+
+
+
+Sometimes, loops do more work than they really *have* to. Take for example, the snippet of code below:
+
+```python
+for i in range(10):
+    x = 3
+    y = y + x
+print(y)
+```
+where what we really want is 
+
+```python
+x = 3
+for i in range(10):
+    y = y + x
+print(y)
+```
+
+The first snippet redundantly sets `x = 3` for *every iteration* when this constant assignment only necessary to do once. This problem scales with larger programs—so how can we remove redundant calculations?
+
+Loop Invariant Code Motion hoists what doesn't need to be in the loop (invariant code) out of the loop. This optimization cuts down the number of instructions executed, by ensuring unnecessary repetition is avoided. Our implementation first identifies movable components, then iteratively moves them. 
+
+The flow of information in our project follows the following visual outline. First we identify natural loops using some existing tools, control flow graphs and dominator trees, which we slightly debugged. These tools were also used to implement reaching definitions. These both then allow us to detect loop invariant instructions to potentially hoist.
+
+Once these instructions are identified, we check if these instructions can be hoisted.
+We move hoisted instructions into a preheader basic block and rename jump targets accordingly.
+After post-processing, we have our new instruction list!
+
+
+<img src="plan.jpeg" style="width: 100%">
+
+
+Skip to the end to optimize your very own `bril` program!
+
+# Loop
+
+All loops considered here are **natural loops**—that is, a cycle with one entry and a **back-edge**. Back-edges are defined as an edge $A \longrightarrow B$ for tail $A$ and head $B$, such that $B$ dominates $A$.  Natural loops are then defined as the smallest set of vertices $L$ with $A,B \in L$ such that for each vertex $v \in L$ we have $v=B$ or PREDS($v$)$\subseteq L$.
+
+In essence, a back-edge is what brings us from the tail of the loop $A$ to the beginning $B$.
+The cycle surrounding this backedge is our loop, where the entrypoint is the start of the cycle.
+If the loop has only one entry, it is a natural loop. Below is an illustration of a natural loop and its back-edge. The natural loop is highlighted, with a labeled back-edge.
+
+<img src="natloop.jpeg" style="width: 70%">
+
+In the case that an edge connected *entry* and *body*, this would no longer be a natural loop.
+ 
+
+### Detecting Natural Loops
+
+
+To find the loop invariant code, first we must detect all natural loops. To accomplish this, we make use of control flow graphs from `cfg.py` and dominator trees from `dom.py` within our three functions. Back-edges are identified with `get_backedges`, and `loopsy` finds the natural loop associated with an input back-edge.
```

Comment:
While I appreciate what you're doing here, I think this might actually be easier to ingest _without_ the snippets of Python. Feel free to link to your code (you can even link to specific line ranges on GitHub), but I think listing the code inline here distracts a little bit from the overview you're trying to give.